README
IO::Pager version 0.30
=====================

IO::Pager - Select a pager and pipe text to it if destination is a TTY


IO::Pager is lightweight and can be used to locate an available pager
and set the PAGER environment variable sanely or as a factory for creating
objects defined elsewhere such as IO::Pager::Buffered and IO::Pager::Unbuffered.

IO::Pager subclasses are designed to programmatically decide whether
or not to pipe a filehandle's output to a program specified in PAGER.
Subclasses may inherit from IO::Pager, and implement the IO handle
methods desired.


INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   # Run non-interactive tests
   make test
   # Run interactive tests
   perl -Mblib t.pl
   make install

DEPENDENCIES

Multiple core modules are used:
base, Env, File::Spec, File::Which, IO::Handle, SelectSaver, Symbol, Tie::Handle

Test::More is used during the installation procedure.

PORTABILITY

IO::Pager is intended to be as portable as possible, I've tested it on numerous
platform and perl version combinations. Most work fine though you may run
into quirks, it tends to be happiest under perl built with perlio not stdio.

COPYRIGHT AND LICENSE

Copyright (C) 2003-2012 Jerrad Pierce

 * Thou shalt not claim ownership of unmodified materials.

 * Thou shalt not claim whole ownership of modified materials.

 * Thou shalt grant the indemnity of the provider of materials.

 * Thou shalt use and dispense freely without other restrictions.

Or, if you prefer:

  This library is free software; you can redistribute it and/or modify it
  under the same terms as Perl itself, either Perl version 5.0 or, at your
  option, any later version of Perl 5 you may have available.


CHANGES
Revision history for Perl extension IO::Pager.

0.35    Fri Jan 01 2016

	Update test 11 for Windows; okay in CMD, PowerShell and MSYS
	but tcsh and win-bash both fail

	Add .travis.yml

0.34	Wed Feb 25 2015

	Fix version error in Makefile.

	Fix a bug in 3-arg open introduced by fix for 1-arg in 0.33.

0.33	Wed Feb 25 2005

	Fix "Warning: unable to close filehandle $real_fh properly."

	Fixed FILEHANDLE only open().

0.32	Tue Feb 24 2005

	Add 3-arg open support.

	Add eof() mechanism for semi-elegant signal handling.

	Update documentation.

	Undef $ENV{LESS} in interactive tests in case -E is set.

	Robustify t/11 for Win32 (RT#79950).

0.31    Sat Apr 06 2013

	Add missing IO::Handle fallback for ::Unbuffered (RT#82351)

	Bogus fix for Windoze automatic line-ending failure of t/11 (RT#79950),
	improvements welcomed.

0.30	Tue Sep 04 2012
	Implement feature request #78270 to include support say

	Preserve layers of passed hilehandle on instantiation.

	Fixed tests under Win32 & cygwin, bug #75181.

	Swap IO::Pager:new and ::open, alter subclass::open to
	remove unnecessary level of subroutine calls

	Return IO::Handle objects when not connected to a TTY for
	transparent use of IO::Pager when programming OO.
	NOTE: This is broken, see ::less in TODO

0.24	Sat Feb 18 2012 
	Fixed bug #74691, tests fail if CPANTS folk don't have a perl in PATH!

	Add PID method & token overload to access pager process identifier.

	Remove semi-circular reference in object to fix implicit close().

0.20	Sat Feb 11 2012

	Fixed bug #74691, add File::Which as dependency / skip test if missing

	Fixed non-TTY/redirection behavior and added corresponding test

	Improve buffering sistuation in both ::Buffered and ::Unbuffered

	Auto-gensym scalar filehandles like perl's open

	Add OO interface

	Add alternate pagers 'lv', 'pg',
	and system default '/etc/alternatives/pager' (bug# 74519)

	Add tests for initialization failure conditions (9-open.t)

0.16	Fri Feb 3 2012

	Require 5.8.0 or better

	Removed Perl warning 'Use of tie on a handle without * is deprecated'.

	No interactive questions during 'make test' (bug# 66718)

	Better handling of missing PAGER.

	Look for more pagers: 'most' and 'w3m' (bug# 74519)

	Documentation clean-up, light refactoring, more error checks.

	Do not run IO::Pager::Page automatically when in Perl compile mode.

0.10	Unreleased

	Added support for encodings via binmode() (bug# 67930).

	Test: encoding, handle (glob, scalar, OO), interactive...

0.07	Unreleased

	Rewrote to support scalar filehandles per user request and IO::Handle
	style OO support came along for the ride.

	Fixed several mis-coded segments that did not affect operation.

	Added C<tell> in IO::Pager::Unbuffered to check size of buffered output.

	Pager open error now sets $! instead of issuing a warning, to more
	closely emulate the C<open> experience. Check return value as before.

	Neglected to update $VERSION in last release.

	Touched up the documentation.

0.06	Wed Sep 21 2005

	Fixed a stupid leftover using =~ for assignment that some perls ignored
	(bug# 14692).

	Fixed other nits in t.pl and documentation.

0.05	Thu Jul 17 20:01:55 EDT 2003

	Touched up documentation.

	Added IO::Pager::Page, a clone of IO::Page "Broken pipe" foible and all.

	Fixed a bug in pager selection algorithm, it failed if you did not
	have File::Which. It is now back to being only strongly recommended as
	opposed to required.

0.04	Wed Jul 16 18:53:58 EDT 2003

	Removed a stupid debug message from IO::Pager::new.

	Improved PAGER selection algorithm.

	Touched up documentation.

	Added some tests.

0.03	Tue Jul 15 23:11:12 2003

	Forked from IO::Page.

0.02 	Tue Jul 15 23:11:12 2003

	original version; created by h2xs 1.22 with options -X -n IO::Pager.


t/TestUtils.pm
package t::TestUtils;

use strict;
use warnings;
use Config;
use Test::More;
use Env qw( HARNESS_ACTIVE );
use ExtUtils::MakeMaker qw( prompt );

use base qw( Exporter );
our @EXPORT;
BEGIN {
   @EXPORT  = qw{
      skip_interactive
      skip_old_perl
      skip_no_file_which
      skip_not_in_path
      is_no
      is_yes
      perl_exe
      perl_path
      prompt
   };
}

sub skip_interactive {
  skip "Run 'perl -Mblib t.pl' to perform interactive tests.", 1 if $HARNESS_ACTIVE;
}

sub skip_old_perl {
  skip "Layers requires Perl 5.8.0 or better.", 1 if $] < 5.008;
}

sub skip_no_file_which {
  skip "This test requires File::Which.", 1 if not eval { require File::Which };
}

sub skip_not_in_path {
  # Test that the specified executable can be found in the PATH environment
  # variable using File::Which.
  my $exe = shift; 
  my $loc = File::Which::which($exe);
  skip "Executable '$exe' is not in PATH.", 1 if not defined $loc;
}

sub is_yes {
  my ($val) = @_;
  return ($val =~ /^y(?:es)?/i || $val eq '');
}

sub is_no {
  my ($val) = @_;
  return ($val =~ /^n(?:o)?/i || $val eq '');
}

sub perl_exe {
  # Find the Perl executable name
  my $this_perl = $^X;
  $this_perl = (File::Spec->splitpath( $this_perl ))[-1];
  return $this_perl;
}

sub perl_path {
  # Find the Perl full-path (taken from the perlvar documentation)
  my $this_perl = $^X;
  if ($^O ne 'VMS') {
    $this_perl .= $Config{_exe}
    unless $this_perl =~ m/$Config{_exe}$/i;
  }
  return $this_perl;
}

1;


t/11-redirect-oo.pl
use blib;
use IO::Pager;
use Test::More;

my $FH = new IO::Pager;

if(! @ARGV){
    our $txt; require 't/08-redirect.pl';
    use Data::Dumper 'Dumper'; print Dumper $FH;
    $FH->print($txt);
}
else{
    isa_ok($FH, 'IO::Pager::less');
    isa_ok($FH, 'IO::Handle');
    done_testing;
}


Makefile.PL
use ExtUtils::MakeMaker;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
    'NAME'              => 'IO::Pager',
    'VERSION_FROM'      => 'lib/IO/Pager.pm', # finds $VERSION
    'BUILD_REQUIRES'    => {
       'Test::More'     => 0,
       'Config'         => 0,
       'Env'            => 0,
       'Exporter'       => 0,
       'File::Temp'     => 0,
    },
    'PREREQ_PM'         => {
       'base'           => 0,
       'Env'            => 0,
       'File::Spec'     => 0,
       'File::Which'    => 0,
       'IO::Handle'     => 0,
       'PerlIO'         => 0,
       'SelectSaver'    => 0,
       'Symbol'         => 0,
       'Tie::Handle'    => 0,
    },
    ($] >= 5.005 ?  # Add these new keywords supported since 5.005
       (
        ABSTRACT_FROM    => 'lib/IO/Pager.pm', # retrieve abstract from module
        AUTHOR           => 'Jerrad Pierce <jpierce@cpan.org>, Florent Angly <florent.angly@gmail.com>',
       ) : ()),
);


lib/IO/Pager/Page.pm
package IO::Pager::Page;

use strict;

our $VERSION = 0.16;


# The meat
BEGIN {
  # Do nothing in Perl compile mode
  return if $^C;
  # Find a pager
  use IO::Pager;
  # Pipe stdout to it
  new IO::Pager *STDOUT, 'IO::Pager::Unbuffered';
}

# Gravy
sub import {
  my ($self, %opt) = @_;
  $SIG{PIPE} = sub{ exit 0; } if $opt{hush};
}

"Badee badee badee that's all folks!";

__END__

=head1 NAME

IO::Pager::Page - Emulate IO::Page, pipe STDOUT to a pager if STDOUT is a TTY

=head1 SYNOPSIS

Pipes STDOUT to a pager if STDOUT is a TTY

=head1 DESCRIPTION

IO::Pager was designed to programmatically decide whether or not to point
the STDOUT file handle into a pipe to program specified in the I<PAGER>
environment variable or one of a standard list of pagers.

=head1 USAGE

  BEGIN {
    use IO::Pager::Page;
    # use I::P::P first, just in case another module sends output to STDOUT
  }
  print<<HEREDOC;
  ...
  A bunch of text later
  HEREDOC

If you wish to forgo the potential for a I<Broken Pipe> foible resulting
from the user exiting the pager prematurely, load IO::Pager::Page like so:

  use IO::Pager::Page hush=>1;

=head1 SEE ALSO

L<IO::Page>, L<IO::Pager>, L<IO::Pager::Unbuffered>, L<IO::Pager::Buffered>

=head1 AUTHOR

Jerrad Pierce <jpierce@cpan.org>

Florent Angly <florent.angly@gmail.com>

This module inspired by Monte Mitzelfelt's IO::Page 0.02

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2003-2012 Jerrad Pierce

=over

=item * Thou shalt not claim ownership of unmodified materials.

=item * Thou shalt not claim whole ownership of modified materials.

=item * Thou shalt grant the indemnity of the provider of materials.

=item * Thou shalt use and dispense freely without other restrictions.

=back

Or, if you prefer:

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.0 or,
at your option, any later version of Perl 5 you may have available.

=cut


lib/IO/Pager/less.pm
package IO::Pager::less;
our $VERSION = 0.30;

use strict;
#use base qw( IO::Handle );
use base qw( IO::Pager );

sub new{#[FH], [Buffered?]
#    warn join(',',@_);
#    my $FH = defined($_[0]) ?
#	IO::Handle->new_from_fd(fileno($_[0]), 'w') :
#	IO::Handle->new->fdopen(Symbol::gensym(), 'w');
#    use Data::Dumper 'Dumper'; print Dumper $FH;
#    $FH->autoflush(1) if $_[1];
##    return bless $FH, 'IO::Pager::less';
#    return $FH;
  my($class, $tied_fh) = @_;
  tie *$tied_fh, $class, $tied_fh or return 0;
}

sub TIEHANDLE{

}

1;

__END__

=head1 NAME

IO::Pager::less - (OO) pagerless output, not output with the less(1) pager

=head1 SYNOPSIS

See L<IO::Pager>

=head1 DESCRIPTION

This is a stub class that is not meant to be invoked directly.
It provides transparent fallback for OO methods when not connected
to a TTY by inherting methods from L<IO::Handle>.

=head1 SEE ALSO

L<IO::Pager::Buffered>,  L<IO::Pager::Unbuffered>

=head1 AUTHOR

Jerrad Pierce <jpierce@cpan.org>

This module was inspired by Monte Mitzelfelt's IO::Page 0.02

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2012 Jerrad Pierce

=over

=item * Thou shalt not claim ownership of unmodified materials.

=item * Thou shalt not claim whole ownership of modified materials.

=item * Thou shalt grant the indemnity of the provider of materials.

=item * Thou shalt use and dispense freely without other restrictions.

=back

Or, if you prefer:

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.0 or,
at your option, any later version of Perl 5 you may have available.

=cut


lib/IO/Pager/Buffered.pm
package IO::Pager::Buffered;
our $VERSION = 0.30;

use strict;
use base qw( IO::Pager );
use SelectSaver;


sub new(;$) {  # [FH]
  my($class, $tied_fh);

  eval { ($class, $tied_fh) = &IO::Pager::_init };
  #We're not on a TTY so...
  if( defined($class) && $class eq '0' or $@ =~ '!TTY' ){
      #...leave filehandle alone if procedural
      return $_[1] if defined($_[2]) && $_[2] eq 'procedural';

      #...fall back to IO::Handle for transparent OO programming
      eval "require IO::Pager::less" or die $@;
      return IO::Pager::less::new($_[1], 0);
  }
  $!=$@, return 0 if $@ =~ 'pipe';

  tie *$tied_fh, $class, $tied_fh or return 0;
}

#Punt to base, preserving FH ($_[0]) for pass by reference to gensym
sub open(;$) { # [FH]
#  IO::Pager::open($_[0], 'IO::Pager::Buffered');
  &new('IO::Pager::Buffered', $_[0]);
}


# Overload IO::Pager methods

sub PRINT {
  my ($self, @args) = @_;
  $self->{buffer} .= join($,||'', @args);
}


sub CLOSE {
  my ($self) = @_;
  # Print buffer and close using IO::Pager's methods
  $self->SUPER::PRINT($self->{buffer}) if exists $self->{buffer};
  $self->SUPER::CLOSE();
}

*DESTROY = \&CLOSE;

sub TELL {
  # Return the size of the buffer
  my ($self) = @_;
  use bytes;
  return exists($self->{buffer}) ? length($self->{buffer}) : 0;
}


sub flush(;*) {
  my ($self) = @_;
  if( exists $self->{buffer} ){
    my $saver = SelectSaver->new($self->{real_fh});
    local $|=1;
    ($_, $self->{buffer}) = ( $self->{buffer}, '');
    $self->SUPER::PRINT($_);
  }
}


1;

__END__

=head1 NAME

IO::Pager::Buffered - Pipe deferred output to PAGER if destination is a TTY

=head1 SYNOPSIS

  use IO::Pager::Buffered;
  {
    local $token = IO::Pager::Buffered::open *STDOUT;
    print <<"  HEREDOC" ;
    ...
    A bunch of text later
    HEREDOC
  }

  {
    # You can also use scalar filehandles...
    my $token = IO::Pager::Buffered::open($FH) or warn($!);
    print $FH "No globs or barewords for us thanks!\n";
  }

  {
    # ...or an object interface
    my $token = new IO::Pager::Buffered;

    $token->print("OO shiny...\n");
  }

=head1 DESCRIPTION

IO::Pager subclasses are designed to programmatically decide whether
or not to pipe a filehandle's output to a program specified in I<PAGER>;
determined and set by IO::Pager at runtime if not yet defined.

This subclass buffers all output for display upon exiting the current scope.
If this is not what you want look at another subclass such as
L<IO::Pager::Unbuffered>. While probably not common, this may be useful in
some cases,such as buffering all output to STDOUT while the process occurs,
showing only warnings on STDERR, then displaying the output to STDOUT after.
Or alternately letting output to STDOUT slide by and defer warnings for later
perusal.

=head1 METHODS

Class-specific method specifics below, others are inherited from IO::Pager.

=head2 open( [FILEHANDLE] )

Instantiate a new IO::Pager to paginate FILEHANDLE if necessary.
I<Assign the return value to a scoped variable>. Output does not
occur until all references to this variable are destroyed eg;
upon leaving the current scope. See L</DESCRIPTION>.

=head2 new( [FILEHANDLE] )

Almost identical to open, except that you will get an L<IO::Handle>
back if there's no TTY to allow for IO::Pager agnostic programming.

=head2 tell( FILEHANDLE )

Returns the size of the buffer in bytes.

=head2 flush( FILEHANDLE )

Immediately flushes the contents of the buffer.

If the last print did not end with a newline, the text from the
preceding newline to the end of the buffer will be flushed but
is unlikely to display until a newline is printed and flushed.

=head1 CAVEATS

If you mix buffered and unbuffered operations the output order is unspecified,
and will probably differ for a TTY vs. a file. See L<perlfunc>.

I<$,> is used see L<perlvar>.

=head1 SEE ALSO

L<IO::Pager>, L<IO::Pager::Unbuffered>, L<IO::Pager::Page>,

=head1 AUTHOR

Jerrad Pierce <jpierce@cpan.org>

Florent Angly <florent.angly@gmail.com>

This module was inspired by Monte Mitzelfelt's IO::Page 0.02

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2003-2012 Jerrad Pierce

=over

=item * Thou shalt not claim ownership of unmodified materials.

=item * Thou shalt not claim whole ownership of modified materials.

=item * Thou shalt grant the indemnity of the provider of materials.

=item * Thou shalt use and dispense freely without other restrictions.

=back

Or, if you prefer:

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.0 or,
at your option, any later version of Perl 5 you may have available.

=cut


lib/IO/Pager/TiedStream.pm
package IO::Pager::TiedStream;

use 5;
use strict;
use Env qw( PAGER );
use base qw( Tie::Handle );

our $VERSION = 0.10;




1;


__END__

=head1 NAME

IO::Pager::TiedStream - 

=head1 SYNOPSIS

  use IO::Pager::TiedStream;
  my $buffered = 0;
  tie *$fh, 'IO::Pager::TiedStream', $fh, $buffered;

=head1 DESCRIPTION

IO::Pager::TiedStream is a class to tie a filehandle and pipe its output
to the pager specified in the PAGER environment variable.

=head1 SEE ALSO

L<IO::Pager>, L<IO::Pager::Unbuffered>, L<IO::Pager::Unbuffered>,
L<IO::Pager::Page>

L<Tie::Handle>

=head1 AUTHOR

Jerrad Pierce <jpierce@cpan.org>

Florent Angly <florent.angly@gmail.com>

This module inspired by Monte Mitzelfelt's IO::Page 0.02

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2003-2012 Jerrad Pierce

=over

=item * Thou shalt not claim ownership of unmodified materials.

=item * Thou shalt not claim whole ownership of modified materials.

=item * Thou shalt grant the indemnity of the provider of materials.

=item * Thou shalt use and dispense freely without other restrictions.

=back

Or, if you prefer:

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.0 or,
at your option, any later version of Perl 5 you may have available.

=cut


lib/IO/Pager.pm
package IO::Pager;
our $VERSION = 0.35; #Really 0.34

use 5.008; #At least, for decent perlio, and other modernisms
use strict;
use base qw( Tie::Handle );
use Env qw( PAGER );
use File::Spec;
use PerlIO;
use Symbol;

use overload '+' => "PID", bool=> "PID";

our $SIGPIPE;

sub find_pager {
  # Return the name (or path) of a pager that IO::Pager can use
  my $io_pager;

  # Use File::Which if available (strongly recommended)
  my $which = eval { require File::Which };

  # Look for pager in PAGER first
  if ($PAGER) {
    # Strip arguments e.g. 'less --quiet'
    my ($pager, @options) = (split ' ', $PAGER);
    $pager = _check_pagers([$pager], $which);
    $io_pager = join ' ', ($pager, @options) if defined $pager;
  }

  # Then search pager amongst usual suspects
  if (not defined $io_pager) {
    my @pagers = ('/etc/alternatives/pager',
		  '/usr/local/bin/less', '/usr/bin/less', '/usr/bin/more');
    $io_pager = _check_pagers(\@pagers, $which) 
  }

  # Then check PATH for other pagers
  if ( (not defined $io_pager) && $which ) {
    my @pagers = ('less', 'most', 'w3m', 'lv', 'pg', 'more');
    $io_pager = _check_pagers(\@pagers, $which );
  }

  # If all else fails, default to more
  $io_pager ||= 'more';

  return $io_pager;
}

sub _check_pagers {
  my ($pagers, $which) = @_;
  # Return the first pager in the list that is usable. For each given pager, 
  # given a pager name, try to finds its full path with File::Which if possible.
  # Given a pager path, verify that it exists.
  my $io_pager = undef;
  for my $pager (@$pagers) {
    # Get full path
    my $loc;
    if ( $which && (not File::Spec->file_name_is_absolute($pager)) ) {
      $loc = File::Which::which($pager);
    } else {
      $loc = $pager;
    }
    # Test that full path is valid (some platforms don't do -x so we use -e)
    if ( defined($loc) && (-e $loc) ) {
      $io_pager = $loc;
      last;
    }
  }
  return $io_pager;
}

#Should have this as first block for clarity, but not with its use of a sub :-/
BEGIN { # Set the $ENV{PAGER} to something reasonable
  $PAGER = find_pager();
}


#Factory
sub open(*;$@) { # FH, [MODE], [CLASS]
  my $args = {procedural=>1};
  $args->{mode} = splice(@_, 1, 1) if scalar(@_) == 3;
  $args->{subclass} = pop if scalar(@_) == 2;
  &new(undef, @_, $args);
}

#Alternate entrance: drop class but leave FH, subclass
sub new(*;$@) { # FH, [MODE], [CLASS]
  shift;

  my %args;
  if( ref($_[-1]) eq 'HASH' ){
    %args = %{pop()};
    warn "REMAINDER? (@_)", scalar @_;
    push(@_, $args{procedural});
  }
  else{
    $args{mode} = splice(@_, 1, 1) if $_[1] =~ /^:/;
    $args{subclass} = pop if exists($_[1]);
  }

  #Leave filehandle in @_ for pass by reference to allow gensym
  $args{subclass} ||= 'IO::Pager::Unbuffered';
  $args{subclass} =~ s/^(?!IO::Pager::)/IO::Pager::/;
  eval "require $args{subclass}" or die "Could not load $args{subclass}: $@\n";
  my $token = $args{subclass}->new(@_);

  if( defined($args{mode}) ){
    $args{mode} =~ s/^\|-//;
    $token->BINMODE($args{mode});
  }
  return $token;
}


sub _init{ # CLASS, [FH] ## Note reversal of order due to CLASS from new()
  #Assign by reference if empty scalar given as filehandle
  $_[1] = gensym() if !defined($_[1]);

  no strict 'refs';
  $_[1] ||= *{select()};

  # Are we on a TTY? STDOUT & STDERR are separately bound
  if ( defined( my $FHn = fileno($_[1]) ) ) {
    if ( $FHn == fileno(STDOUT) ) {
      die '!TTY' unless -t $_[1];
    }
    if ( $FHn == fileno(STDERR) ) {
      die '!TTY' unless -t $_[1];
    }
  }

  #XXX This allows us to have multiple pseudo-STDOUT
  #return 0 unless -t STDOUT;

  return ($_[0], $_[1]);
}


# Methods required for implementing a tied filehandle class

sub TIEHANDLE {
  my ($class, $tied_fh) = @_;
  unless ( $PAGER ){
    die "The PAGER environment variable is not defined, you may need to set it manually.";
  }
  my($real_fh, $child);
  if ( $child = CORE::open($real_fh, '|-', $PAGER) ){
    my @oLayers = PerlIO::get_layers($tied_fh, details=>1, output=>1);
    my $layers = '';
    for(my $i=0;$i<$#oLayers;$i+=3){
      #An extra base layer requires more keystrokes to exit
      next if $oLayers[$i] =~ /unix|stdio/ && !defined($oLayers[+1]);

      $layers .= ":$oLayers[$i]";
      $layers .=  '(' . ($oLayers[$i+1]) . ')' if defined($oLayers[$i+1]);
    }
    CORE::binmode($real_fh, $layers);
  }
  else{
    die "Could not pipe to PAGER ('$PAGER'): $!\n";
  }
  return bless {
                'real_fh' => $real_fh,
                'child'   => $child,
		'pager'   => $PAGER,
               }, $class;
}


sub BINMODE {
  my ($self, $layer) = @_;
  CORE::binmode($self->{real_fh}, $layer||':raw');
}

sub WNOHANG();
sub EOF {
  my $self = shift;

  unless( defined($SIGPIPE) ){
    eval 'use POSIX ":sys_wait_h";';
    $SIGPIPE = 0;
  }

  $SIG{PIPE} = sub { $SIGPIPE = 1 unless $ENV{IP_EOF};
		     CORE::close($self->{real_fh});
		     waitpid($self->{child}, WNOHANG);
		     CORE::open($self->{real_fh}, '>&1'); };
  return $SIGPIPE;
}


sub PRINT {
  my ($self, @args) = @_;
  CORE::print {$self->{real_fh}} @args or die "Could not print to PAGER: $!\n";
}

sub PRINTF {
  my ($self, $format, @args) = @_;
  $self->PRINT(sprintf($format, @args));
}

sub say {
  my ($self, @args) = @_;
  $args[-1] .= "\n";
  $self->PRINT(@args);
}

sub WRITE {
  my ($self, $scalar, $length, $offset) = @_;
  $self->PRINT(substr($scalar, $offset||0, $length));
}


sub TELL {
  #Buffered classes provide their own, and others may use this in another way
  return undef;
}


sub CLOSE {
  my ($self) = @_;
  CORE::close($self->{real_fh});
}

*DESTROY = \&CLOSE;


#Non-IO methods
sub PID{
  my ($self) = @_;
  return $self->{child};
}


#Provide lowercase aliases for accessors
foreach my $method ( qw(BINMODE CLOSE EOF PRINT PRINTF TELL WRITE PID) ){
  no strict 'refs';
  *{lc($method)} = \&{$method};
}


1;

__END__

=head1 NAME

IO::Pager - Select a pager and pipe text to it if destination is a TTY

=head1 SYNOPSIS

  # Select an appropriate pager and set the PAGER environment variable
  use IO::Pager;

  # TIMTOWTDI Object-oriented
  {
    # open()                           # Use all the defaults.
    my $object = new IO::Pager;

    # open FILEHANDLE                  # Unbuffered is default subclass
    my $object = new IO::Pager *STDOUT;

    # open FILEHANDLE,EXPR             # Specify subclass
    my $object = new IO::Pager *STDOUT,  'Unbuffered';

    # Direct subclass instantiation    # FH is optional
    use IO::Pager::Unbuffered;
    my $object = new IO::Pager::Unbuffered  *STDOUT;


    $object->print("OO shiny...\n");
    print "Some other text sent to STODUT, perhaps from a foreign routine."

    # $object passes out of scope and filehandle is automagically closed
  }

  # TIMTOWTDI Procedural
  {
    # open FILEHANDLE                    # Unbuffered is default subclass
    my $token = IO::Pager::open *STDOUT;

    # open FILEHANDLE,EXPR               # Specify subclass
    my $token = IO::Pager::open *STDOUT,  'Unbuffered';

    # open FILEHANDLE,MODE,EXPR          # En lieu of a separate binmode()
    my $token = IO::Pager::open *STDOUT, '|-:utf8', 'Unbuffered';


    print <<"  HEREDOC" ;
    ...
    A bunch of text later
    HEREDOC

    # $token passes out of scope and filehandle is automagically closed
  }

  {
    # You can also use scalar filehandles...
    my $token = IO::Pager::open(my $FH) or warn($!); XXX
    print $FH "No globs or barewords for us thanks!\n";
  }


=head1 DESCRIPTION

IO::Pager can be used to locate an available pager and set the I<PAGER>
environment variable (see L</NOTES>). It is also a factory for creating
I/O objects such as L<IO::Pager::Buffered> and L<IO::Pager::Unbuffered>.

IO::Pager subclasses are designed to programmatically decide whether
or not to pipe a filehandle's output to a program specified in I<PAGER>.
Subclasses may implement only the IO handle methods desired and inherit
the remainder of those outlined below from IO::Pager. For anything else,
YMMV. See the appropriate subclass for implementation specific details.

=head1 METHODS

=head2 new( FILEHANDLE, [MODE], [SUBCLASS] )

Almost identical to open, except that you will get an L<IO::Handle>
back if there's no TTY to allow for IO::Pager-agnostic programming.

=head2 open( FILEHANDLE, [MODE], [SUBCLASS] )

Instantiate a new IO::Pager, which will paginate output sent to
FILEHANDLE if interacting with a TTY.

Save the return value to check for errors, use as an object,
or for implict close of OO handles when the variable passes out of scope.

=over

=item FILEHANDLE

You may provide a glob or scalar.

Defaults to currently select()-ed F<FILEHANDLE>.

=item SUBCLASS

Specifies which variety of IO::Pager to create.
This accepts fully qualified packages I<IO::Pager::Buffered>,
or simply the third portion of the package name I<Buffered> for brevity.

Defaults to L<IO::Pager::Unbuffered>.

Returns false and sets I<$!> on failure, same as perl's C<open>.

=back

=head2 PID

Call this method on the token returned by C<open> to get the process
identifier for the child process i.e; pager; if you need to perform
some long term process management e.g; perl's C<waitpid>

You can also access the PID by numifying the instantiation token like so:

  my $child = $token+0;

=head2 close( FILEHANDLE )

Explicitly close the filehandle, this stops any redirection of output
on FILEHANDLE that may have been warranted.

I<This does not default to the current filehandle>.

Alternatively, you may rely upon the implicit close of lexical handles
as they pass out of scope e.g;

  {
     IO::Pager::open local *RIBBIT;
     print RIBBIT "No toad sexing allowed";
     ...
  }
  #The filehandle is closed to additional output

  {
     my $token = new IO::Pager::Buffered;
     $token->print("I like trains");
     ...
  }
  #The string "I like trains" is flushed to the pager, and the handle closed

=head2 binmode( FILEHANDLE, [LAYER] )

Used to set the I/O layer a.k.a. discipline of a filehandle,
such as C<':utf8'> for UTF-8 encoding.

=head2 eof( FILEHANDLE )

Used in the eval-until-eof idiom below, I<IO::Pager> will handle broken pipes
from deceased children for you in one of two ways. If I<$ENV{IP_EOF}> is
false then program flow will pass out of the loop on I<SIGPIPE>, this is the
default. If the variable is true, then the program continues running with
output for the previously paged filehandle directed to the I<STDOUT> stream;
more accurately, the filehandle is reopened to file descriptor 1.

  use IO::Pager::Page; #or whichever you prefer;
  ...
  eval{
    say "Producing prodigious portions of product";
    ...
  } until( eof(*STDOUT) );
  print "Cleaning up after our child before terminating."

If using eof() with L<less>, especially when IP_EOF is set, you may want to
use the I<--no-init> option by setting I<$ENV{IP_EOF}='X'> to prevent the
paged output from being erased when the pager exits.

=head2 print( FILEHANDLE LIST )

print() to the filehandle.

=head2 printf( FILEHANDLE FORMAT, LIST )

printf() to the filehandle.

=head2 syswrite( FILEHANDLE, SCALAR, [LENGTH], [OFFSET] )

syswrite() to the filehandle.

=head1 ENVIRONMENT

=over

=item IP_EOF

Controls IO:Pager behavior when C<eof> is used.

=item PAGER

The location of the default pager.

=item PATH

If the location in PAGER is not absolute, PATH may be searched.

See L</NOTES> for more information.

=back

=head1 FILES

IO::Pager may fall back to these binaries in order if I<PAGER> is not
executable.

=over

=item /etc/alternatives/pager

=item /usr/local/bin/less

=item /usr/bin/less

=item /usr/bin/more

=back

See L</NOTES> for more information.

=head1 NOTES

The algorithm for determining which pager to use is as follows:

=over

=item 1. Defer to I<PAGER>

If the I<PAGER> environment variable is set, use the pager it identifies,
unless this pager is not available.

=item 2. Usual suspects

Try the standard, hardcoded paths in L</FILES>.

=item 3. File::Which

If File::Which is available, use the first pager possible amongst
C<less>, C<most>, C<w3m>, C<lv>, C<pg> and L<more>.

=item 4. more

Set I<PAGER> to C<more>, and cross our fingers.

=back

Steps 1, 3 and 4 rely upon the I<PATH> environment variable.

=head1 SEE ALSO

L<IO::Pager::Buffered>, L<IO::Pager::Unbuffered>, L<IO::Pager::Page>,

L<IO::Page>, L<Meta::Tool::Less>

=head1 AUTHOR

Jerrad Pierce <jpierce@cpan.org>

Florent Angly <florent.angly@gmail.com>

This module was inspired by Monte Mitzelfelt's IO::Page 0.02

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2003-2015 Jerrad Pierce

=over

=item * Thou shalt not claim ownership of unmodified materials.

=item * Thou shalt not claim whole ownership of modified materials.

=item * Thou shalt grant the indemnity of the provider of materials.

=item * Thou shalt use and dispense freely without other restrictions.

=back

Or, if you prefer:

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.0 or,
at your option, any later version of Perl 5 you may have available.

=cut
